var/spack/repos/builtin/packages/mptensor/package.py

```diff
@@ -68,33 +68,43 @@ def cmake_args(self):
     def setup_build_tests(self):
         """Copy the build test files after the package is installed to an
         install test subdirectory for use during `spack test run`."""
+
+        # Tests only supported when spec built with mpi
+        if "+mpi" not in self.spec:
+            print("Skipping copy of stand-alone test files: requires +mpi build")
+            return
+
         self.cache_extra_test_sources(".")
 
-    def test(self):
+        # Clean cached makefiles now so only done once
+        print("Converting cached Makefile for stand-alone test use")
+        with working_dir(join_path(self.install_test_root, "tests"), create=False):
+            make("clean")
+            makefile = FileFilter("Makefile")
+            makefile.filter("g++", "{0}".format(spack_cxx), string=True)
+
+        print("Converting cached Makefile.option for stand-alone test use")
+        with working_dir(join_path(install_test_root(self)), create=False):
+            makefile = FileFilter("Makefile.option")
+            makefile.filter("CXX =.*", "CXX ={0}".format(self.spec["mpi"].mpicxx))
+            makefile.filter("CXXFLAGS =.*", "CXXFLAGS ={0}".format(self.compiler.cxx11_flag))
+
+    def test_mpi(self):
+        """test with +mpi"""
         if "+mpi" not in self.spec:
-            print("Test of mptensor only runs with +mpi option.")
-        else:
-            with working_dir(join_path(self.install_test_root, "tests"), create=False):
-                make("clean")
-                makefile = FileFilter("Makefile")
-                makefile.filter("g++", "{0}".format(spack_cxx), string=True)
-
-            with working_dir(join_path(self.install_test_root), create=False):
-                makefile = FileFilter("Makefile.option")
-                makefile.filter("CXX =.*", "CXX ={0}".format(self.spec["mpi"].mpicxx))
-                makefile.filter("CXXFLAGS =.*", "CXXFLAGS ={0}".format(self.compiler.cxx11_flag))
-
-            math_libs = (
-                self.spec["scalapack"].libs + self.spec["lapack"].libs + self.spec["blas"].libs
-            )
+            raise SkipTest("Package must be installed with +mpi")
+
+        math_libs = self.spec["scalapack"].libs + self.spec["lapack"].libs + self.spec["blas"].libs
 
-            with working_dir(join_path(self.install_test_root, "tests"), create=False):
-                make("LDFLAGS={0}".format(math_libs.ld_flags))
+        with working_dir(self.test_suite.current_test_cache_dir.tests):
+            make = which("make")
+            make(f"LDFLAGS={math_libs.ld_flags}")
 
-                mpirun = self.spec["mpi"].prefix.bin.mpirun
-                mpiexec = Executable(mpirun)
-                mpiexec("-n", "1", "tensor_test.out")
+            mpirun = self.spec["mpi"].prefix.bin.mpirun
+            # mpiexec = Executable(mpirun)
+            mpiexec = which(mpirun)
+            mpiexec("-n", "1", "tensor_test.out")
```

Comment:
At a minimum remove the comment; however, having `mpirun` and `mpiexec` can be "confusing"/"misleading".  Better to have the variable name match the executable (e.g., for `spack pkg grep`).  Suggest:

```suggestion
            mpirun = which(self.spec["mpi"].prefix.bin.mpirun)
            mpirun("-n", "1", "tensor_test.out")
```